File: src/python_ros2_basics/python_ros2_basics/subscriber_example.py
#!/usr/bin/env python3

# ros2 python library
import rclpy
# The node class from rclpy
from rclpy.node import Node
# The message that we are using
from std_msgs.msg import String

# Create a class that inherit from ros2 rclpy Node
class TextSubscriber(Node):
    def __init__(self):
        # Use the init of the Node class
        super().__init__('text_subscriber')
        # Declare a subscriber that will subscribe to the topic with String message type
        # example_string_topic, with the callbach self.listener_callback and a queue size of 10
        self.subscription = self.create_subscription(
            String, 
            "example_string_topic", 
            self.listener_callback, 
            10)
        # Create a class attribute to store the message from topic
        self.last_message = String()
        # Print the init log of the node
        self.get_logger().info('Subscriber node has been started.')

    
    # The callback associtated to the subscriber
    def listener_callback(self, msg):
        self.last_message.data = msg.data
        self.get_logger().info(f'Received: {self.last_message.data}')


# The main function executed when we are launching the node
def main(args=None):
    # Init rclpy ROS2
    rclpy.init(args=args)
    # Create the node
    node = TextSubscriber()
    # Keep the node running
    rclpy.spin(node)
    # Destroy the node on stop
    node.destroy_node()
    # Shutdown the rclpy ROS2
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/cpp_ros2_basics/src/subscriber_example.cpp
// Include necessary C++ headers
#include <memory>
// ROS2 C++ Client library
#include "rclcpp/rclcpp.hpp"
// Include the message type that we will subscribe
#include "std_msgs/msg/string.hpp"

// Define the TextSubscriber class, inheriting from rclcpp::Node
class TextSubscriber : public rclcpp::Node {
public:
    TextSubscriber() : Node("text_subscriber") {
        // Initialize the subscriber, subscribing to example_string_topic with a queue size of 10
        subscription_ = this->create_subscription<std_msgs::msg::String>(
            "example_string_topic", 
            10, 
            std::bind(&TextSubscriber::listener_callback, this, std::placeholders::_1)
        );

        // Log a message indicating the subscriber node has started
        RCLCPP_INFO(this->get_logger(), "Subscriber node has been started.");
    }

private:
    // Callback function called when a message is received
    void listener_callback(const std_msgs::msg::String::SharedPtr msg) {
        last_message.data = msg->data;
        // Log the received message
        RCLCPP_INFO(this->get_logger(), "Received: '%s'", last_message.data.c_str());
    }

    // Member variable for the subscriber
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
    std_msgs::msg::String last_message;
};

// Main function, similar to Python main
int main(int argc, char * argv[]) {
    // Initialize rclcpp
    rclcpp::init(argc, argv);

    // Create a shared pointer for the TextSubscriber node
    auto node = std::make_shared<TextSubscriber>();

    // Keep the node spinning to process incoming messages
    rclcpp::spin(node);

    // Shutdown and cleanup
    rclcpp::shutdown();
    return 0;
}
